service/history/workflow_task_handler_callbacks.go

```diff
@@ -393,6 +396,22 @@ func (handler *workflowTaskHandlerCallbacksImpl) handleWorkflowTaskCompleted(
 	if err != nil {
 		return nil, err
 	}
+	weContext := workflowContext.GetContext()
+	ms := workflowContext.GetMutableState()
+
+	currentWorkflowTask := ms.GetWorkflowTaskByID(token.GetScheduledEventId())
+	if !ms.IsWorkflowExecutionRunning() ||
+		currentWorkflowTask == nil ||
+		currentWorkflowTask.StartedEventID == common.EmptyEventID ||
+		(token.StartedEventId != common.EmptyEventID && token.StartedEventId != currentWorkflowTask.StartedEventID) ||
+		(token.StartTimeUnixnano != 0 && token.StartTimeUnixnano != currentWorkflowTask.StartedTime.UnixNano()) ||
+		currentWorkflowTask.Attempt != token.Attempt ||
+		(token.Version != common.EmptyVersion && token.Version != currentWorkflowTask.Version) {
+		// we have not alter mutable state yet, so release with it with nil to avoid clear MS.
+		workflowContext.GetReleaseFn()(nil)
+		return nil, serviceerror.NewNotFound("Workflow task not found.")
```

Comment:
It probably worth to add what values were expected and what what were actual values in the token, to error message. Otherwise it could be very hard to debug this error.

Reply:
There are so many things in the token that we checked, do you want to add all of them?

Reply:
`tag.Value()` or `tag.NewAnyTag()` can dump all of those (exported) field at once I think.

Reply:
It is error returned from the func, not log.